### ISSUE TITLE
Memberships

### ISSUE THREAD

**Issue by ericgio** [yekrojam/api#35]
In exploring the one-off case more deeply, I'm running into some questions and potential issues around memberships and how they should be handled. Capturing here for discussion and to think about in relation to the multi-tenant case.

1. **Create user and membership at the same time**
When an org adds (creates) a new user, the membership should be generated as well.
2. **Memberships should be unique**
It's currently possible to create multiple membership documents for the same org/user pair. That potentially clutters the DB and creates issues around which is the "correct" membership. Could we do an "update or create" here?
3. **Including roles with user data**
I haven't found a great way to include the roles along with a user's data. The best way I found so far is to do a query like: `/membership?populate=user&query={"user":<userId>,"org":<orgId>}`. Then on the client I get that user object and monkey-patch the roles onto it. Supposedly I can use `$lookup` to somehow join the user and membership documents, but I haven't been able to do that successfully. In any case, it doesn't seem great to push that kind of querying logic to the client.
4. **Org-specific user querying**
Closely related to the above, I'm unaware of an easy way to get all the users for a given org. Right now I'm doing something like the memberships query above and transforming the results into an array of users with the roles attached. It'd be much more intuitive to be able to do something like `/user?query={"org":<orgId>}`.

**Comment by makinde** [comment]
We chatted a bit about (1) and (2) at Ola's

For (3) and (4), this is a bit more about what I was saying at Ola's. The concept of "users for an org" doesn't really exist in the API, and it forces you to be more explicit about what you want. `User`s are connected to `Org`s in multiple ways, and that relationship is represented by the `Membership`. So you might want to find out who the admins of an org are, or who the members are. But you have to be specific when you ask. In the future, there might be other types/roles of membership as well. Perhaps invitees, or people who've requested to join. Because there are lots of potential relationships (and even with just two right now), the API forces you to explicitly ask for what you want.

I'm guessing when you say "users for an org" above, you're wondering how to get the people who are `MEMBER`s of the org. The query would be this (assuming you want to users to display): `/membership?query={"role":"MEMBER","org":<orgId>}&populate=user`. Then if you wanted to display all the users, you could use lodash to extract the user objects for display. -or- you could even have your display widget display the membership object (in case it wants to display if the person is an admin or something). The `Membership` is a first class object that represent a relationship to the org.

Perhaps there's something tricky here about the way the data is stored on the client side as well. I'm not 100% how you'd want to model these objects in the client side store.

One question we also briefly discussed was how to know what the current user can do. I think that was one use case you had for trying to merge the `roles` information into the user object. Again, I'm not 100% sure how the client data modeling/storage works, but from the little I know, I'd advise storing the logged in User's `Membership` as a first class object, and referencing that for the the things you need to figure out. [Side note] If you want to know if the logged in user can edit the org, once the permissions stuff is hooked up, the object will come back with a `permissions` field that will tell you which fields the person can write to on the Org object. If that array is empty, they can't write to any fields, so you could decide not to link them to the admin page.

**Comment by ericgio** [comment]
Yes, I'm currently doing exactly what you suggested (query the membership endpoint). I guess what I'm suggesting is that the API could make this much easier for clients and avoid all the extra logic and transformations. For example, maybe there's a separate `/member` endpoint that requires the org id and returns an array of users for that org, and includes their roles/permissions in each object.

> I'd advise storing the logged in User's Membership as a first class object

I've gone down the route of storing edges on the client in the past and it can get pretty complex and messy. Basically, you end up doing joins on the client all over the place. Multiply that by the number of clients and it's a lot of extra logic that in my opinion should live on the API.

Might be best to discuss in person, since it seems we still have some fundamental disagreements over where logic like this belongs.

**Comment by makinde** [comment]
Yeah, I think our misalignment is around what is the primary unit being dealt with. A User or a Membership. I think the difference there probably stems from Single- vs Multi-tenant perspective. In the single tenant case, there's the temptation to collapse User & Membership concepts into one. In the Multi-tenant case, those two concepts must be kept very distinct (and the primary thing the app cares about is, "what is this user's relationship to the org". You're viewing the User as the top level object to be handled, while I'm viewing the membership, not as an edge, but as the primary top level object to handle.

Not suggesting any specific action/direction here. Rather, trying to highlight what may be the root of what we're talking about. Just sent out some time to chat for next week on FB.

**Comment by makinde** [comment]
Side note, as we talk about this and I learn more, I'm a bit surprised that folks (it may have been you, but I'm pretty sure I've heard it from others) that GraphQL removes the need to a lot of custom endpoints based on what data the client needs. It seems that lots of custom work is needed both with REST (and the ability to populate) and GraphQL...that is, if the client wants to interact with the API having it roll up a lot of the logic depending on how the client wants to see it).

**Comment by makinde** [comment]
http://collabedit.com/jtdjy
